### PR TITLE
Better math in code comments

### PR DESCRIPTION
**Pull request by tzanio**
Added a filter to allow us to use regular LaTeX math formulas (`$`...`$` and `$$`...`$$`) instead of the Doxygen style (`\f$`, `\f[`, `\f]`, etc.)

See e.g. the description of class Hybridization:

Screenshot 2024-01-31 at 2 56 07 PM

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4103](https://github.com/mfem/mfem/pull/4103) | @tzanio | @tzanio | @vladotomov + @pazner + @termi-official + @v-dobrev + @najlkin | 2/2/24 | 2/19/24 | ⌛due 2/19/24 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

File: doc/CodeDocumentation.conf.in [hunk 1]
```diff
@@ -1125,7 +1125,13 @@ IMAGE_PATH             =
 # need to set EXTENSION_MAPPING for the extension otherwise the files are not
 # properly processed by doxygen.
 
-INPUT_FILTER           =
+# Allow the use of standard LaTeX formulas in MFEM comments by replacing
+# $$...$$ with \f[...\f] and $...$ with \f$...\f$ before running Doxygen.
+# The refgular expression ((?:.|\n)+?) is a lazy match for one or more
+# characters, include newline. If presetn, Doxygen-style LaTeX commands
```
Comment (pazner) [inline doc/CodeDocumentation.conf.in:1131]:
```suggestion
# characters, include newline. If present, Doxygen-style LaTeX commands
```

File: doc/CodeDocumentation.conf.in [hunk 2]
```diff
@@ -1799,7 +1805,7 @@ MATHJAX_VERSION        = MathJax_2
 # The default value is: HTML-CSS.
 # This tag requires that the tag USE_MATHJAX is set to YES.
 
-MATHJAX_FORMAT         = HTML-CSS
+MATHJAX_FORMAT         = SVG
```
Comment (pazner) [inline doc/CodeDocumentation.conf.in:1808]: Is SVG better?

File: doc/CodeDocumentation.conf.in [hunk 3]
```diff
@@ -1125,7 +1125,13 @@ IMAGE_PATH             =
 # need to set EXTENSION_MAPPING for the extension otherwise the files are not
 # properly processed by doxygen.
 
-INPUT_FILTER           =
+# Allow the use of standard LaTeX formulas in MFEM comments by replacing
+# $$...$$ with \f[...\f] and $...$ with \f$...\f$ before running Doxygen.
```
Comment (pazner) [inline doc/CodeDocumentation.conf.in:1129]: Could we have it replace `\[` and `\]` with `\f[` and `\f]` also?

**Review by pazner — COMMENTED** [review]
This is great, it's a huge improvement to readability!

File: doc/CodeDocumentation.conf.in (hunk 2 above)
Comment (tzanio) [reply doc/CodeDocumentation.conf.in:1808]: For me, the rendering with `HTML-CSS` did not look good (on Mac/Chrome). Will be good to get data from other folks.

File: doc/CodeDocumentation.conf.in (hunk 3 above)
Comment (tzanio) [reply doc/CodeDocumentation.conf.in:1129]: Certainly

**Comment by najlkin** [comment]
I agree, that is what I meant basically and with that in-built parameter of doxygen it is pretty straightforward :wink:

File: doc/CodeDocumentation.conf.in (hunk 2 above)
Comment (v-dobrev) [reply doc/CodeDocumentation.conf.in:1808]: Does this mean SVGs are generated by `doxygen` for every formula, or that the SVGs are generated only in-browser by mathjax to render the formulas? If the former is true, then I think we should revert this -- having a ton of `.svg` files for every formula seems quite excessive.

It looks like this may control the "MathJax output format": https://docs.mathjax.org/en/latest/output/index.html, so SVGs are probably generated on-the-fly in the browser.

Maybe we can tweak some other MathJax option (e.g. the math font, if that's part of the doxygen-generated html) to get nice looking "HTML-CSS". Or, maybe we can try MathJax v3 instead of v2, see https://www.doxygen.nl/manual/config.html#cfg_use_mathjax. According to these docs, "HTML-CSS (which is slower, but has the best compatibility. ...)" so I guess that's why it's the default.

**Comment by v-dobrev** [comment]
This is definitely an improvement over having the `\f`s everywhere.

File: doc/CodeDocumentation.conf.in (hunk 2 above)
Comment (tzanio) [reply doc/CodeDocumentation.conf.in:1808]: This is something I will appreciate some help with (to make sure it is not due to my setup/browser/OS).

If you have time, can you try the different MathJax output formats and see if you can use `MathJax_3`?

For me `HTML-CSS` did not look good and `MathJax_3` didn't work on a first try.

Comment (tzanio) [reply doc/CodeDocumentation.conf.in:1808]: Here's what `HTML-CSS` looks like for me (compare with the image in the PR description)

Screenshot 2024-02-01 at 8 33 56 AM

Comment (tzanio) [reply doc/CodeDocumentation.conf.in:1808]: Here is what `MATHJAX_VERSION = MathJax_3` plus `MATHJAX_FORMAT = chtml` looks like:

Comment (tzanio) [reply doc/CodeDocumentation.conf.in]: Switched to MathJax3 in 44d33d17f32377ab9fbea204b57e5aa16d0a40d5 -- thanks @najlkin !

**Comment by najlkin** [comment]
Unfortunately, MathJax 3 was introduced in [Doxygen 1.9.2](https://www.doxygen.nl/manual/changelog.html#log_1_9_2), so it fails on Ubuntu with Doxygen 1.9.1 😔

File: doc/CodeDocumentation.conf.in (hunk 3 above)
Comment (tzanio) [reply doc/CodeDocumentation.conf.in:1129]: If we want to add this, the regex pattern will be `s/\\\[((?:.|\n)+?)\\\]/\\f[\1\\f]/g;`

**Comment by tzanio** [comment]
> Unfortunately, MathJax 3 was introduced in [Doxygen 1.9.2](https://www.doxygen.nl/manual/changelog.html#log_1_9_2), so it fails on Ubuntu with Doxygen 1.9.1 😔

We switched the config files to Doxygen 1.9.8 recently

**Comment by najlkin** [comment]
> > Unfortunately, MathJax 3 was introduced in [Doxygen 1.9.2](https://www.doxygen.nl/manual/changelog.html#log_1_9_2), so it fails on Ubuntu with Doxygen 1.9.1 😔
> 
> We switched the config files to Doxygen 1.9.8 recently

But when you look into the log of the test, it says: "Setting up doxygen (1.9.1-2ubuntu2) ...". Maybe we could switch between the versions of MathJax in the make file depending on what is installed? In the worst case, there could be two doxygen configuration files 🤔

**Comment by najlkin** [comment]
Ok, it passes now 🙃 , because doxygen just gives a warning about the unknown parameter `MATHJAX_VERSION` and not an error. In reality, the resulting HTML documents are broken due to that `MATHJAX_EXTENSIONS`, but the test does not check that. So if we require the users to have MathJax 3 and only keep this "compatibility" with MathJax 2 for the test, it is fine. It is ugly, but works 😄

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Comment by termi-official** [comment]
Not a doxygen expert, but is this really valid to do? I thought $ is reserved by doxygen for variables. I cannot find the details in the manual, but a hint is given here https://www.doxygen.nl/manual/commands.html#cmddollar .

**Comment by tzanio** [comment]
> Not a doxygen expert, but is this really valid to do? I thought $ is reserved by doxygen for variables. I cannot find the details in the manual, but a hint is given here https://www.doxygen.nl/manual/commands.html#cmddollar .

Doxygen hopefully will not see those because it will first apply the filter which replaces them?

File: doc/CodeDocumentation.conf.in [hunk 4]
```diff
@@ -1125,7 +1125,13 @@ IMAGE_PATH             =
...
+# The refgular expression ((?:.|\n)+?) is a lazy match for one or more
```
Comment (termi-official) [inline doc/CodeDocumentation.conf.in:1130]:
```suggestion
# The regular expression ((?:.|\n)+?) is a lazy match for one or more
```

**Comment by termi-official** [comment]
Exactly, I am not sure whether if the filter will also gobble up doxygen variables (if there are at least 2 in one line).

**Comment by najlkin** [comment]
> Exactly, I am not sure whether if the filter will also gobble up doxygen variables (if there are at least 2 in one line).

I have quickly searched that (`[^f]\$`) and we do not use it in the code at the moment, so a conceptual transgression, but not that bad to my taste 😁

**Comment by tzanio** [comment]
Added a [few commits](https://github.com/mfem/mfem/compare/7d2b5b8..doxy-math-dev) with the following changes:

- Mentioned the math-in-comments rules in `CONTRIBUTING`
- Check for old-style comments in GitHub CI as part of `repo-check`

I did not add support for `\[...]\` as @pazner [requested](https://github.com/mfem/mfem/pull/4103#discussion_r1475131081) because that may make it more challenging to render e.g. in the VSCode [plugin](https://marketplace.visualstudio.com/items?itemName=Remisa.mathover), but we can do that if the majority prefers it.

From my perspective, this is good to go 🚢

File: doc/CodeDocumentation.conf.in [hunk 5]
```diff
@@ -1125,7 +1125,13 @@ IMAGE_PATH             =
...
 # properly processed by doxygen.
 
-INPUT_FILTER           =
+# Allow the use of standard LaTeX formulas in MFEM comments by replacing
+# $$...$$ with \f[...\f] and $...$ with \f$...\f$ before running Doxygen.
+# The regular expression ((?:.|\n)+?) is a lazy match for one or more
+# characters, include newline. If present, Doxygen-style LaTeX commands
+# such as \f$, \f[, etc., are left unchanged.
+
+INPUT_FILTER           = perl -0777 -pe 's/\$\$((?:.|\n)+?)\$\$/\\f[\1\\f]/g; s/(?<!\\f)\$((?:.|\n)+?)\$/\\f\$\1\\f\$/g'
```
[the remark this answers is not in the document]
Comment (tzanio) [reply doc/CodeDocumentation.conf.in:1134]: I'm not sure all of the above (e.g. `(?<!\\f)`) is supported in `sed`. If it is, we can switch.

[the remark this answers is not in the document]
Comment (najlkin) [reply doc/CodeDocumentation.conf.in:1134]: Right, sed does not support many of those more advanced constructions. There are some workarounds for some (or maybe all?) of them. I played with that a while, but it would be too convoluted, so I am giving up 😏 Doxygen is not an essential part of the library, so who wants must pay the little toll of installing `perl`.

**Review by najlkin — APPROVED** [review]

File: doc/CodeDocumentation.conf.in (hunk 5 above)
[the remark this answers is not in the document]
Comment (tzanio) [reply doc/CodeDocumentation.conf.in:1134]: At least on Linux and Mac perl is typically pre-installed.

**Review by vladotomov — APPROVED** [review]
Huge improvement, thank you!

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

File: doc/CodeDocumentation.conf.in (hunk 5 above)
Comment (v-dobrev) [inline doc/CodeDocumentation.conf.in:1134]: When I use an older `doxygen` (1.8.14) on macOS, I'm getting errors like these:
```console
...
Reading /Users/dobrev1/code/mfem/mfem.hpp...
sh: perl-0777-pes/\$\$((?:.|\n)+?)\$\$/\\f[\1\\f]/g;s/(?<!\\f)\$((?:.|\n)+?)\$/\\f\$\1\\f\$/g: No such file or directory
Parsing file /Users/dobrev1/code/mfem/mfem.hpp...
...
```
Putting double-quotes (`"`) around the command seems to fix this issue:
```
INPUT_FILTER           = "perl -0777 -pe 's/\$\$((?:.|\n)+?)\$\$/\\f[\1\\f]/g; s/(?<!\\f)\$((?:.|\n)+?)\$/\\f\$\1\\f\$/g'"
```
I'm not sure, but I assume newer `doxygen` doesn't have that issue on macOS.

Comment (v-dobrev) [reply doc/CodeDocumentation.conf.in:1134]: Also, with the older `doxygen` version, the math formulas are not displayed properly. If I revert the changes to the MATHJAX options, the math is fine.

Comment (tzanio) [reply doc/CodeDocumentation.conf.in:1134]: We can add the quotes, but overall I think it is OK to switch to the new Doxygen version -- the documentation is mostly build by us, and that requirement is [documented](https://github.com/mfem/mfem/blob/master/doc/README#L17)

Comment (v-dobrev) [reply doc/CodeDocumentation.conf.in:1134]: See #4149.
